pkg/common_controller/snapshot_controller.go

```diff
@@ -166,27 +183,74 @@ func (ctrl *csiSnapshotController) syncContent(content *crdv1.VolumeSnapshotCont
 // created, updated or periodically synced. We do not differentiate between
 // these events.
 // For easier readability, it is split into syncUnreadySnapshot and syncReadySnapshot
-func (ctrl *csiSnapshotController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
-	klog.V(5).Infof("synchonizing VolumeSnapshot[%s]: %s", snapshotKey(snapshot), getSnapshotStatusForLogging(snapshot))
+func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
+	klog.V(5).Infof("synchronizing VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
+
+	// If content is deleted already, remove SnapshotBound finalizer
+	content, err := ctrl.contentExists(snapshot)
+	if err != nil {
+		return err
+	}
+	deleteContent := false
+	// It is possible for contentExists to return nil, nil
+	if content != nil && content.Spec.DeletionPolicy == crdv1.VolumeSnapshotContentDelete {
+		klog.V(5).Infof("syncSnapshot: Content [%s] deletion policy [%s] is delete.", content.Name, content.Spec.DeletionPolicy)
+		deleteContent = true
+	}
+
+	snapshotBound := false
+	// Check if the snapshot content is bound to the snapshot
+	if content != nil && utils.IsSnapshotBound(snapshot, content) {
+		klog.Infof("syncSnapshot: VolumeSnapshot %s is bound to volumeSnapshotContent [%s]", snapshot.Name, content.Name)
+		snapshotBound = true
+	}
 
-	if isSnapshotDeletionCandidate(snapshot) {
+	// Check is snapshot deletionTimestamp is set and any finalizer is on
+	if utils.IsSnapshotDeletionCandidate(snapshot) {
 		// Volume snapshot should be deleted. Check if it's used
 		// and remove finalizer if it's not.
 		// Check if a volume is being created from snapshot.
-		isUsed := ctrl.isVolumeBeingCreatedFromSnapshot(snapshot)
-		if !isUsed {
-			klog.V(5).Infof("syncSnapshot: Remove Finalizer for VolumeSnapshot[%s]", snapshotKey(snapshot))
-			return ctrl.removeSnapshotFinalizer(snapshot)
+		inUse := ctrl.isVolumeBeingCreatedFromSnapshot(snapshot)
+
+		klog.V(5).Infof("syncSnapshot[%s]: set DeletionTimeStamp on content.", utils.SnapshotKey(snapshot))
+		// If content exists, set DeletionTimeStamp on the content;
+		// content won't be deleted immediately due to the finalizer
+		if content != nil && deleteContent {
+			err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().Delete(*snapshot.Status.BoundVolumeSnapshotContentName, &metav1.DeleteOptions{})
```

Comment:
what if *snapshot.Status is nil (status is lost case?),

Reply:
Actually I don't need to rely on snapshot.Status here.  We have content so I can use content.Name directly.